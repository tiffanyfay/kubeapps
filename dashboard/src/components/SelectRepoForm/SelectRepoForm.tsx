import * as React from "react";
import { Link } from "react-router-dom";

import { IAppRepository, IRBACRole } from "../../shared/types";
import LoadingWrapper from "../LoadingWrapper";

import { ErrorSelector, MessageAlert } from "../ErrorAlert";

interface ISelectRepoFormProps {
  isFetching: boolean;
  namespace: string;
  kubeappsNamespace: string;
  repoError?: Error;
  error?: Error;
  repo: IAppRepository;
  repos: IAppRepository[];
  chartName: string;
  checkChart: (namespace: string, repo: string, chartName: string) => any;
  fetchRepositories: (namespace: string) => void;
}

interface ISelectRepoFormState {
  repo: string;
}

class SelectRepoForm extends React.Component<ISelectRepoFormProps, ISelectRepoFormState> {
  public state: ISelectRepoFormState = {
    repo:
      this.props.repo.metadata && this.props.repo.metadata.name
        ? this.props.repo.metadata.name
        : "",
  };

  public componentDidMount() {
    this.props.fetchRepositories(this.props.namespace);
  }

  public render() {
    if (this.props.isFetching) {
      return <LoadingWrapper />;
    }
    if (this.props.repoError) {
      return (
        <ErrorSelector
          error={this.props.repoError}
          namespace={this.props.kubeappsNamespace}
          action="view"
          defaultRequiredRBACRoles={{ view: this.requiredRBACRoles() }}
          resource="App Repositories"
        />
      );
    }
    if (this.props.repos.length === 0) {
      return (
        <MessageAlert
          level={"warning"}
          children={
            <div>
              <h5>Chart repositories not found.</h5>
              Manage your Helm chart repositories in Kubeapps by visiting the{" "}
              <Link to={`/config/ns/${this.props.namespace}/repos`}>
                App repositories configuration
              </Link>{" "}
              page.
            </div>
          }
        />
      );
    }
    return (
      <LoadingWrapper loaded={this.props.repos.length > 0}>
        <div className="container margin-normal">
          <div className="col-8">
            {this.props.error && (
              <ErrorSelector
                error={this.props.error}
                resource={`Chart ${this.state.repo}/${this.props.chartName}`}
              />
            )}
          </div>
          <div className="col-12">
            <h2>Select the source repository of {this.props.chartName}</h2>
          </div>
          <div className="col-8">
            <label htmlFor="chartRepoName">Chart Repository Name *</label>
            <select
              id="chartRepoName"
              onChange={this.handleChartRepoNameChange}
              value={this.state.repo}
              required={true}
            >
              {!this.state.repo && <option key="" value="" />}
              {this.props.repos.map(r => (
                <option key={r.metadata.name} value={r.metadata.name}>
                  {r.metadata.name} ({this.getRepoURL(r.metadata.name)})
                </option>
              ))}
            </select>
          </div>
          <div>
            <p>
              {" "}
              * If the repository containing {this.props.chartName} is not in the list add it{" "}
              <a href="/config/repos"> here </a>.{" "}
            </p>
          </div>
        </div>
      </LoadingWrapper>
    );
  }

  public handleChartRepoNameChange = async (e: React.ChangeEvent<HTMLSelectElement>) => {
    this.props.checkChart(this.props.namespace, e.currentTarget.value, this.props.chartName);
    this.setState({ repo: e.currentTarget.value });
  };

  private getRepoURL = (name: string) => {
    let res = "";
    this.props.repos.forEach(r => {
      if (r.metadata.name === name && r.spec) {
        res = r.spec.url;
      }
    });
    return res;
  };

  private requiredRBACRoles(): IRBACRole[] {
    return [
      {
        apiGroup: "kubeapps.com",
        namespace: this.props.kubeappsNamespace,
        resource: "apprepositories",
        verbs: ["get"],
      },
    ];
  }
}

export default SelectRepoForm;
