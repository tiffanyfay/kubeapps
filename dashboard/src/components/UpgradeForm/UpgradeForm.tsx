import { RouterAction } from "connected-react-router";
import * as jsonpatch from "fast-json-patch";
import { JSONSchema4 } from "json-schema";
import * as React from "react";
import * as YAML from "yaml";

import { deleteValue, setValue } from "../../shared/schema";
import { IChartState, IChartVersion } from "../../shared/types";
import DeploymentFormBody from "../DeploymentFormBody/DeploymentFormBody";
import { ErrorSelector } from "../ErrorAlert";
import LoadingWrapper from "../LoadingWrapper";

export interface IUpgradeFormProps {
  appCurrentVersion: string;
  appCurrentValues?: string;
  chartName: string;
  namespace: string;
  releaseName: string;
  repo: string;
  repoNamespace: string;
  error?: Error;
  selected: IChartState["selected"];
  deployed: IChartState["deployed"];
  upgradeApp: (
    version: IChartVersion,
    releaseName: string,
    namespace: string,
    values?: string,
    schema?: JSONSchema4,
  ) => Promise<boolean>;
  push: (location: string) => RouterAction;
  goBack: () => RouterAction;
  fetchChartVersions: (namespace: string, id: string) => Promise<IChartVersion[]>;
  getChartVersion: (namespace: string, id: string, chartVersion: string) => void;
}

interface IUpgradeFormState {
  appValues: string;
  valuesModified: boolean;
  isDeploying: boolean;
  modifications?: jsonpatch.Operation[];
}

class UpgradeForm extends React.Component<IUpgradeFormProps, IUpgradeFormState> {
  public state: IUpgradeFormState = {
    appValues: this.props.appCurrentValues || "",
    isDeploying: false,
    valuesModified: false,
  };

  public componentDidMount() {
    const chartID = `${this.props.repo}/${this.props.chartName}`;
    this.props.fetchChartVersions(this.props.repoNamespace, chartID);
  }

  public componentDidUpdate = (prevProps: IUpgradeFormProps) => {
    let modifications = this.state.modifications;
    if (this.props.deployed.values && !modifications) {
      // Calculate modifications from the default values
      const defaultValuesObj = YAML.parse(this.props.deployed.values);
      const deployedValuesObj = YAML.parse(this.props.appCurrentValues || "");
      modifications = jsonpatch.compare(defaultValuesObj, deployedValuesObj);
      this.setState({ modifications });
      this.setState({ appValues: this.applyModifications(modifications, this.state.appValues) });
    }

    if (prevProps.selected.version !== this.props.selected.version && !this.state.valuesModified) {
      // Apply modifications to the new selected version
      const appValues = modifications
        ? this.applyModifications(modifications, this.props.selected.values || "")
        : this.props.selected.values || "";
      this.setState({ appValues });
    }
  };

  public render() {
    const { namespace, releaseName, error, selected } = this.props;
    if (error) {
      return (
        <ErrorSelector error={error} namespace={namespace} action="update" resource={releaseName} />
      );
    }
    if (selected.versions.length === 0) {
      return <LoadingWrapper />;
    }
    const chartID = `${this.props.repo}/${this.props.chartName}`;
    return (
      <form className="container padding-b-bigger" onSubmit={this.handleDeploy}>
        <div className="row">
          <div className="col-12">
            <h2>{`${this.props.releaseName} (${chartID})`}</h2>
          </div>
          <div className="col-8">
            <DeploymentFormBody
              chartNamespace={this.props.repoNamespace}
              chartID={chartID}
              chartVersion={this.props.appCurrentVersion}
              deployedValues={this.applyModifications(
                this.state.modifications || [],
                this.props.deployed.values || "",
              )}
              namespace={this.props.namespace}
              releaseVersion={this.props.appCurrentVersion}
              selected={this.props.selected}
              push={this.props.push}
              goBack={this.props.goBack}
              getChartVersion={this.props.getChartVersion}
              setValues={this.handleValuesChange}
              appValues={this.state.appValues}
              setValuesModified={this.setValuesModified}
            />
          </div>
        </div>
      </form>
    );
  }

  public setValuesModified = () => {
    this.setState({ valuesModified: true });
  };

  public handleValuesChange = (value: string) => {
    this.setState({ appValues: value });
  };

  public handleDeploy = async (e: React.FormEvent<HTMLFormElement>) => {
    e.preventDefault();
    const { selected, push, upgradeApp, releaseName, namespace } = this.props;
    const { appValues } = this.state;

    this.setState({ isDeploying: true });
    if (selected.version) {
      const deployed = await upgradeApp(
        selected.version,
        releaseName,
        namespace,
        appValues,
        selected.schema,
      );
      this.setState({ isDeploying: false });
      if (deployed) {
        push(`/apps/ns/${namespace}/${releaseName}`);
      }
    }
  };

  private applyModifications(modifications: jsonpatch.Operation[], appValues: string) {
    // And we add any possible change made to the original version
    if (modifications.length) {
      modifications.forEach(modification => {
        // Transform the JSON Path to the format expected by setValue
        // /a/b/c => a.b.c
        const path = modification.path.replace(/^\//, "").replace(/\//g, ".");
        if (modification.op === "remove") {
          appValues = deleteValue(appValues, path);
        } else {
          // Transform the modification as a ReplaceOperation to read its value
          const value = (modification as jsonpatch.ReplaceOperation<any>).value;
          appValues = setValue(appValues, path, value);
        }
      });
    }
    return appValues;
  }
}

export default UpgradeForm;
