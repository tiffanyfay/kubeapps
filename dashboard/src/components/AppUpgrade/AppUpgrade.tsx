import * as React from "react";

import { RouterAction } from "connected-react-router";
import { JSONSchema4 } from "json-schema";
import { IAppRepository, IChartState, IChartVersion, IRelease } from "../../shared/types";
import { ErrorSelector } from "../ErrorAlert";
import LoadingWrapper from "../LoadingWrapper";
import SelectRepoForm from "../SelectRepoForm";
import UpgradeForm from "../UpgradeForm";

interface IAppUpgradeProps {
  app: IRelease;
  appsIsFetching: boolean;
  appsError: Error | undefined;
  namespace: string;
  releaseName: string;
  repoName: string;
  repoNamespace: string;
  selected: IChartState["selected"];
  deployed: IChartState["deployed"];
  upgradeApp: (
    version: IChartVersion,
    releaseName: string,
    namespace: string,
    values?: string,
    schema?: JSONSchema4,
  ) => Promise<boolean>;
  fetchChartVersions: (namespace: string, id: string) => Promise<IChartVersion[]>;
  getAppWithUpdateInfo: (namespace: string, releaseName: string) => void;
  getChartVersion: (namespace: string, id: string, chartVersion: string) => void;
  getDeployedChartVersion: (namespace: string, id: string, chartVersion: string) => void;
  push: (location: string) => RouterAction;
  goBack: () => RouterAction;
  // repo selector properties
  reposIsFetching: boolean;
  kubeappsNamespace: string;
  repoError?: Error;
  chartsError: Error | undefined;
  repo: IAppRepository;
  repos: IAppRepository[];
  checkChart: (namespace: string, repo: string, chartName: string) => any;
  fetchRepositories: (namespace: string) => void;
}

class AppUpgrade extends React.Component<IAppUpgradeProps> {
  public componentDidMount() {
    const { releaseName, getAppWithUpdateInfo, namespace } = this.props;
    getAppWithUpdateInfo(namespace, releaseName);
  }

  public componentDidUpdate(prevProps: IAppUpgradeProps) {
    const { app, repoName, repoNamespace } = this.props;
    if (app && repoName) {
      const { chart } = app;
      if (
        chart &&
        chart.metadata &&
        chart.metadata.name &&
        chart.metadata.version &&
        (prevProps.app !== app || prevProps.repoName !== repoName)
      ) {
        const chartID = `${repoName}/${chart.metadata.name}`;
        this.props.getDeployedChartVersion(repoNamespace, chartID, chart.metadata.version);
      }
    }
  }

  public render() {
    const {
      app,
      namespace,
      appsError,
      releaseName,
      appsIsFetching,
      repoName,
      repoNamespace,
      selected,
      deployed,
      upgradeApp,
      push,
      goBack,
      fetchChartVersions,
      getChartVersion,
    } = this.props;
    if (appsError) {
      return (
        <ErrorSelector
          error={appsError}
          namespace={namespace}
          action="update"
          resource={releaseName}
        />
      );
    }
    if (appsIsFetching || !app || !app.updateInfo) {
      return <LoadingWrapper />;
    }
    const repo = repoName || app.updateInfo.repository.name;
    if (app && app.chart && app.chart.metadata && repo) {
      return (
        <div>
          <UpgradeForm
            appCurrentVersion={app.chart.metadata.version!}
            appCurrentValues={(app.config && app.config.raw) || ""}
            chartName={app.chart.metadata.name!}
            repo={repo}
            repoNamespace={repoNamespace}
            namespace={namespace}
            releaseName={releaseName}
            selected={selected}
            deployed={deployed}
            upgradeApp={upgradeApp}
            push={push}
            goBack={goBack}
            fetchChartVersions={fetchChartVersions}
            getChartVersion={getChartVersion}
          />
        </div>
      );
    }

    return (
      <SelectRepoForm
        isFetching={this.props.reposIsFetching}
        error={this.props.chartsError}
        kubeappsNamespace={this.props.kubeappsNamespace}
        namespace={this.props.namespace}
        repoError={this.props.repoError}
        repo={this.props.repo}
        repos={this.props.repos}
        chartName={app.chart?.metadata?.name!}
        checkChart={this.props.checkChart}
        fetchRepositories={this.props.fetchRepositories}
      />
    );
  }
}

export default AppUpgrade;
