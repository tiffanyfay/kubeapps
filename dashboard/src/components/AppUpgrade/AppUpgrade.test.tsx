import { mount, shallow } from "enzyme";
import context from "jest-plugin-context";
import * as React from "react";
import { BrowserRouter } from "react-router-dom";

import { hapi } from "shared/hapi/release";
import itBehavesLike from "../../shared/specs";
import { ForbiddenError, IAppRepository, IChartState, IRelease } from "../../shared/types";
import { ErrorSelector, MessageAlert, PermissionsErrorAlert } from "../ErrorAlert";
import ErrorPageHeader from "../ErrorAlert/ErrorAlertHeader";
import SelectRepoForm from "../SelectRepoForm";
import UpgradeForm from "../UpgradeForm";
import AppUpgrade from "./AppUpgrade";

const defaultProps = {
  app: {} as hapi.release.Release,
  appsIsFetching: false,
  reposIsFetching: false,
  repoName: "",
  repoNamespace: "chart-namespace",
  isFetching: false,
  checkChart: jest.fn(),
  clearRepo: jest.fn(),
  appsError: undefined,
  chartsError: undefined,
  fetchChartVersions: jest.fn(),
  fetchRepositories: jest.fn(),
  getAppWithUpdateInfo: jest.fn(),
  getChartVersion: jest.fn(),
  deployed: {} as IChartState["deployed"],
  getDeployedChartVersion: jest.fn(),
  kubeappsNamespace: "kubeapps",
  namespace: "default",
  push: jest.fn(),
  goBack: jest.fn(),
  releaseName: "foo",
  repo: {} as IAppRepository,
  repoError: undefined,
  repos: [],
  selected: {} as IChartState["selected"],
  upgradeApp: jest.fn(),
  version: "1.0.0",
};

beforeEach(() => {
  jest.resetAllMocks();
});

itBehavesLike("aLoadingComponent", {
  component: AppUpgrade,
  props: { ...defaultProps, isFetching: true },
});

it("renders the repo selection form if not introduced", () => {
  const wrapper = shallow(
    <AppUpgrade
      {...defaultProps}
      app={
        {
          chart: {
            metadata: {
              name: "bar",
              version: "1.0.0",
            },
          },
          name: "foo",
          updateInfo: { repository: {} },
        } as IRelease
      }
      repos={[
        {
          metadata: { name: "stable" },
        } as IAppRepository,
      ]}
    />,
  );
  expect(wrapper.find(SelectRepoForm)).toExist();
  expect(wrapper.find(ErrorSelector)).not.toExist();
  expect(wrapper.find(UpgradeForm)).not.toExist();
  expect(wrapper).toMatchSnapshot();
});

context("when an error exists", () => {
  it("renders a generic error message", () => {
    const repo = {
      metadata: { name: "stable" },
    } as IAppRepository;
    const wrapper = shallow(
      <AppUpgrade
        {...defaultProps}
        appsError={new Error("foo doesn't exists")}
        repos={[repo]}
        repo={repo}
      />,
    );

    expect(wrapper.find(ErrorSelector)).toExist();
    expect(wrapper.find(SelectRepoForm)).not.toExist();
    expect(wrapper.find(UpgradeForm)).not.toExist();

    expect(wrapper.html()).toContain("Sorry! Something went wrong.");
    expect(wrapper).toMatchSnapshot();
  });

  it("renders a forbidden message", () => {
    const role = {
      apiGroup: "kubeapps.com",
      namespace: "kubeapps",
      resource: "apprepositories",
      verbs: ["get"],
    };
    const wrapper = mount(
      <AppUpgrade
        {...defaultProps}
        app={
          {
            chart: {
              metadata: {
                name: "bar",
                version: "1.0.0",
              },
            },
            name: "foo",
            updateInfo: { repository: {} },
          } as IRelease
        }
        repoError={new ForbiddenError(JSON.stringify([role]))}
      />,
    );
    expect(wrapper.find(ErrorSelector)).toExist();
    expect(wrapper.find(UpgradeForm)).not.toExist();

    expect(wrapper.find(ErrorPageHeader).text()).toContain(
      "You don't have sufficient permissions to view App Repositories in the kubeapps namespace",
    );
    expect(wrapper.find(PermissionsErrorAlert).prop("roles")[0]).toMatchObject(role);
  });

  it("renders a warning message if there are no repositories", () => {
    const wrapper = mount(
      <BrowserRouter>
        <AppUpgrade
          {...defaultProps}
          app={
            {
              chart: {
                metadata: {
                  name: "bar",
                  version: "1.0.0",
                },
              },
              name: "foo",
              updateInfo: { repository: {} },
            } as IRelease
          }
          repos={[]}
        />
      </BrowserRouter>,
    );

    expect(wrapper.find(SelectRepoForm).find(MessageAlert)).toExist();
    expect(wrapper.find(UpgradeForm)).not.toExist();

    expect(
      wrapper
        .find(MessageAlert)
        .children()
        .text(),
    ).toContain("Chart repositories not found");
  });
});

it("renders the upgrade form when the repo is available", () => {
  const wrapper = shallow(
    <AppUpgrade
      {...defaultProps}
      app={
        {
          chart: {
            metadata: {
              name: "bar",
              version: "1.0.0",
            },
          },
          name: "foo",
          updateInfo: { repository: {} },
        } as IRelease
      }
    />,
  );
  wrapper.setProps({ repoName: "foobar" });
  expect(wrapper.find(UpgradeForm)).toExist();
  expect(wrapper.find(ErrorSelector)).not.toExist();
  expect(wrapper.find(SelectRepoForm)).not.toExist();
  expect(wrapper).toMatchSnapshot();
});

it("skips the repo selection form if the app contains upgrade info", () => {
  const repo = {
    metadata: { name: "stable" },
  } as IAppRepository;
  const app = {
    chart: {
      metadata: {
        name: "bar",
        version: "1.0.0",
      },
    },
    name: "foo",
    updateInfo: {
      upToDate: true,
      chartLatestVersion: "1.1.0",
      appLatestVersion: "1.1.0",
      repository: { name: "stable", url: "" },
    },
  } as IRelease;
  const wrapper = shallow(<AppUpgrade {...defaultProps} repos={[repo]} />);
  wrapper.setProps({ app });
  expect(wrapper.find(UpgradeForm)).toExist();
  expect(wrapper.find(ErrorSelector)).not.toExist();
  expect(wrapper.find(SelectRepoForm)).not.toExist();
  expect(wrapper).toMatchSnapshot();
});

describe("when receiving new props", () => {
  it("should request the deployed chart when the app and repo are populated", () => {
    const app = {
      chart: {
        metadata: {
          name: "bar",
          version: "1.0.0",
        },
      },
    } as IRelease;
    const getDeployedChartVersion = jest.fn();
    const wrapper = shallow(
      <AppUpgrade {...defaultProps} getDeployedChartVersion={getDeployedChartVersion} />,
    );
    wrapper.setProps({ repoName: "stable", app });
    wrapper.update();
    expect(getDeployedChartVersion).toHaveBeenCalledWith(defaultProps.repoNamespace, "stable/bar", "1.0.0");
  });

  it("should request the deployed chart when the repo is populated later", () => {
    const app = {
      chart: {
        metadata: {
          name: "bar",
          version: "1.0.0",
        },
      },
    } as IRelease;
    const getDeployedChartVersion = jest.fn();
    const wrapper = shallow(
      <AppUpgrade {...defaultProps} app={app} getDeployedChartVersion={getDeployedChartVersion} />,
    );
    expect(getDeployedChartVersion).not.toHaveBeenCalled();
    wrapper.setProps({ repoName: "stable" });
    wrapper.update();
    expect(getDeployedChartVersion).toHaveBeenCalledWith(defaultProps.repoNamespace, "stable/bar", "1.0.0");
  });

  it("a new app should re-trigger the deployed chart retrieval", () => {
    const app = {
      chart: {
        metadata: {
          name: "bar",
          version: "1.0.0",
        },
      },
    } as IRelease;
    const getDeployedChartVersion = jest.fn();
    const wrapper = shallow(
      <AppUpgrade {...defaultProps} getDeployedChartVersion={getDeployedChartVersion} />,
    );
    wrapper.setProps({ repoName: "stable", app });
    expect(getDeployedChartVersion).toHaveBeenCalledWith(defaultProps.repoNamespace, "stable/bar", "1.0.0");

    const app2 = {
      chart: {
        metadata: {
          name: "foobar",
          version: "1.0.0",
        },
      },
    } as IRelease;
    wrapper.setProps({ app: app2 });
    expect(getDeployedChartVersion).toHaveBeenCalledWith(defaultProps.repoNamespace, "stable/foobar", "1.0.0");
  });
});
