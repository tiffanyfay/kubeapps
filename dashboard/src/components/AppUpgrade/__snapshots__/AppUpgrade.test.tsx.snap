// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`loading spinner matches the snapshot 1`] = `
<LoadingWrapper
  loaded={false}
  type={0}
/>
`;

exports[`renders the repo selection form if not introduced 1`] = `
<SelectRepoForm
  chartName="bar"
  checkChart={[Function]}
  fetchRepositories={[Function]}
  isFetching={false}
  kubeappsNamespace="kubeapps"
  namespace="default"
  repo={Object {}}
  repos={
    Array [
      Object {
        "metadata": Object {
          "name": "stable",
        },
      },
    ]
  }
/>
`;

exports[`renders the upgrade form when the repo is available 1`] = `
<div>
  <UpgradeForm
    appCurrentValues=""
    appCurrentVersion="1.0.0"
    chartName="bar"
    deployed={Object {}}
    fetchChartVersions={[Function]}
    getChartVersion={[Function]}
    goBack={[Function]}
    namespace="default"
    push={[Function]}
    releaseName="foo"
    repo="foobar"
    repoNamespace="chart-namespace"
    selected={Object {}}
    upgradeApp={[Function]}
  />
</div>
`;

exports[`skips the repo selection form if the app contains upgrade info 1`] = `
<div>
  <UpgradeForm
    appCurrentValues=""
    appCurrentVersion="1.0.0"
    chartName="bar"
    deployed={Object {}}
    fetchChartVersions={[Function]}
    getChartVersion={[Function]}
    goBack={[Function]}
    namespace="default"
    push={[Function]}
    releaseName="foo"
    repo="stable"
    repoNamespace="chart-namespace"
    selected={Object {}}
    upgradeApp={[Function]}
  />
</div>
`;

exports[`when an error exists renders a generic error message 1`] = `
<ErrorSelector
  action="update"
  error={[Error: foo doesn't exists]}
  namespace="default"
  resource="foo"
/>
`;
